### PR TITLE
PixelCNN in Linen API

### PR DESCRIPTION
**Pull request by marcvanzee**
With lots of help from @j-towns, thanks!

Side-by-side diff of change: https://github.com/marcvanzee/flax/compare/linen2..pixelcnn2

### CODE REVIEW

File: linen_examples/pixelcnn/pixelcnn.py [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+        (k_h - 1, 0),  # Vertical padding
+        (k_w // 2, k_w // 2))  # Horizontal padding
+
+    return Conv(
+        self.features,
+        self.kernel_size,
+        self.strides,
+        padding,
+        init_scale=self.init_scale)(
+            inputs)
```
[the remark this answers is not in the document]
Comment (marcvanzee) [reply linen_examples/pixelcnn/pixelcnn.py:268]: I agree.. I just ran it through Google's Python linter and this is what it came up with :-).

**Comment by avital** [comment]
@marcvanzee It'd be great to see a side-by-side diff of the model changes. Could you please push a separate remote branch where the changes are made directly to examples/pixelcnn, and then share a GitHub diff URL?

File: linen_examples/pixelcnn/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+### Requirements (Sampling)
+* [Pillow](https://pillow.readthedocs.io/en/stable/) for saving samples as PNG files.
+
+### Supported setups
+The model should run with other configurations and hardware, but was tested on the following.
+
+| Hardware | Batch size | Training time | Log-likelihood (bits/dimension) | TensorBoard.dev |
+| --- | --- | --- | --- | --- |
+| 8 x TPUv3  | 256  |   8h 55m | 3.03 | [2020-08-13](https://tensorboard.dev/experiment/XpXXKlOxRjmu8P2z49rPAA/) |
```
[the remark this answers is not in the document]
Comment (marcvanzee) [reply linen_examples/pixelcnn/README.md:24]: Yes, I agree having a fair comparison with TPU is nice. I actually didn't know 2.92 was an important point to reach, thanks for clarifying. I'm now rerunning on TPU and I'll add the numbers once it i done.

File: linen_examples/pixelcnn/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+### How to run
+#### 8 x TPUv3
+To run training
+```
+python train.py --batch_size=320
```
[the remark this answers is not in the document]
Comment (marcvanzee) [reply linen_examples/pixelcnn/README.md:30]: I've now changed it to 320 since TPUv3 has the same RAM as the Nvidia GPU V100.

File: linen_examples/pixelcnn/pixelcnn.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcvanzee) [reply linen_examples/pixelcnn/pixelcnn.py:268]: I've now reduced the line breaking.

File: linen_examples/pixelcnn/train.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+      images, means, inv_scales, logit_weights)
+  return -jnp.mean(log_likelihoods) / (jnp.log(2) * jnp.prod(images.shape[-3:]))
+
+
+def train_step(optimizer, ema, batch, learning_rate_fn, dropout_rng=None):
+  """Perform a single training step."""
+
+  def loss_fn(params):
+    """loss function used for training."""
+    logits = model(dropout_p=FLAGS.dropout_rate).apply(
```
[the remark this answers is not in the document]
Comment (marcvanzee) [reply linen_examples/pixelcnn/train.py:131]: Changed it.

File: linen_examples/pixelcnn/train.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+  initial_variables = model().init({
+      'param': init_rng,
+      'dropout': dropout_rng
+  }, init_batch)['param']
+  optimizer_def = optim.Adam(
+      learning_rate=FLAGS.learning_rate, beta1=0.95, beta2=0.9995)
+  optimizer = optimizer_def.create(initial_variables)
+
+  # TODO(marcvanzee): Restoring checkpoints is broken. Fix this.
```
[the remark this answers is not in the document]
Comment (marcvanzee) [reply linen_examples/pixelcnn/train.py:221]: Yes I agree fixing it would be good. But since the example works fine for now, it does not seem blocking for this PR. I'll looking into it, and once it is fixed I will uncomment this. WDYT?

File: linen_examples/pixelcnn/pixelcnn.py [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+
+from flax import linen as nn
+
+
+# High level model definition
+class PixelCNNPP(nn.Module):
+  """High-level model definition."""
+  depth: int = 5
+  features: int = 160
+  k: int = 10
```
[the remark this answers is not in the document]
Comment (marcvanzee) [reply linen_examples/pixelcnn/pixelcnn.py:55]: I agree, changed.

File: linen_examples/pixelcnn/pixelcnn.py [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+    return inputs + a * nn.sigmoid(b)
+
+
+ResDown = partial(GatedResnet, conv_module=ConvDown)
+ResDownRight = partial(GatedResnet, conv_module=ConvDownRight)
+
+
+# Logistic mixture distribution utils
+def conditional_params_from_outputs(theta, img):
+  """Maps an image `img` and the PixelCNN++ convnet output `theta` to conditional parameters for a mixture of k logistics over each pixel.
```
[the remark this answers is not in the document]
Comment (marcvanzee) [reply linen_examples/pixelcnn/pixelcnn.py:357]: Done.

**Comment by marcvanzee** [comment]
> @marcvanzee It'd be great to see a side-by-side diff of the model changes. Could you please push a separate remote branch where the changes are made directly to examples/pixelcnn, and then share a GitHub diff URL?

Added a side-by-side diff in the PR description.

File: linen_examples/pixelcnn/train.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (j-towns) [reply linen_examples/pixelcnn/train.py:221]: Yeah, agreed, if the problem is unrelated to PixelCNN I don't think it's worth blocking on. I thought @levskaya made it sound as though it wouldn't be too difficult to fix though :).
